### ISSUE TITLE
Try removing the `Full` constructor

### ISSUE THREAD

**Issue by sjakobi** [haskell-unordered-containers/unordered-containers#399]
In my mind `Full` nodes should be pretty rare in the real world. The root node might be one, but at deeper levels I wouldn't expect to encounter any `Full` nodes.

Therefore it's unclear what benefit we get from all the dedicated code operating on this type of node.

From an [earlier experiment](https://github.com/haskell-unordered-containers/unordered-containers/issues/75#issuecomment-463844448) to remove `Full` we have a report of perf regressions, but it's not clear which code regressed and why.

I would expect that the current `Int` benchmarks show regressions due to their unusually contiguous hash distribution, which I think is rather unrealistic.

**Comment by treeowl** [comment]
Yeah, my experiments were never very complete. Would surely be worth trying again. As I recall, I didn't really go all the way changing all the code. I think I tried making `Full` a pattern synonym and just worked through actually blasting it in a few places. That could have been part of the problem, or not. Changing all the code is a lot of work, but I agree with you that it doesn't seem likely to occur anywhere other than the root for non-`Int`-like instances.

**Comment by sjakobi** [comment]
@treeowl You mentioned that you wanted to take another stab at removing `Full` in https://github.com/haskell-unordered-containers/unordered-containers/issues/445#issuecomment-3351637446.

I agree that this is worth doing, primarily because it would remove so much code that we'd otherwise have to continue to maintain, and because it would simplify future API enhancements and other changes.

But since the top priority for this package is performance, I wonder how we can quantify the performance benefits of the reduced code size.

For example, I'd expect more inlining to happen. Do you have a plan on how to measure the benefits of this, @treeowl?

On the other hand I'd like to get a good idea of the performance impact of replacing `Full` nodes with `BitmapIndexed` nodes on functions like `lookup`, `insert`, etc. IMO the current benchmark suite is fairly useless for this. It's hard to get reliable timings from it and they nearly all are for operations on hashmaps of ~4000 elements. Therefore I plan on overhauling the benchmark suite fairly soon.
